ast/sql.go

```diff
@@ -621,38 +491,18 @@ func (a *ArrayLiteral) SQL() string {
 }
 
 func (s *StructLiteral) SQL() string {
-	sql := "STRUCT"
-	if s.Fields != nil {
-		sql += "<"
-		for i, f := range s.Fields {
-			if i != 0 {
-				sql += ", "
-			}
-			sql += f.SQL()
-		}
-		sql += ">"
-	}
-	sql += "("
-	for i, v := range s.Values {
-		if i != 0 {
-			sql += ", "
-		}
-		sql += v.SQL()
-	}
-	sql += ")"
-	return sql
+	// TODO: len(s.Fields) > 0 is better than s.Fields != nil, but it need to update current testdata.
```

Comment:
We intentionally distinct `STRUCT` and `STRUCT<>` because the first means a struct having any fields, and the second is the struct with no fields. For instance, `select CAST(struct(1) AS STRUCT<x INT64>).x` is valid, but `select CAST(struct<>(1) AS STRUCT<x INT64>).x` is invalid.

Reply:
Thank you for your explanation.
I understood your intention.

> We intentionally distinct STRUCT and STRUCT<> because the first means a struct having any fields, and the second is the struct with no fields. For instance, select `CAST(struct(1) AS STRUCT<x INT64>).x` is valid, but select `CAST(struct<>(1) AS STRUCT<x INT64>).x` is invalid.

My thought is that:
* It is `StructLiteral`, not `StructType`, so there is no semantic difference between `STRUCT<>()` and `STRUCT()`. It can be normalized without ambiguity.
* `STRUCT<>()` is valid, but it can feel unfamiliar because it has no fields to name or type.
* The official docs describe only [`STRUCT()`](https://cloud.google.com/spanner/docs/reference/standard-sql/data-types#typeless_struct_syntax).

It might be a good idea to comment that both `STRUCT()` and `STRUCT<>()` are lexically preserved.

Reply:
Perhaps, we need to split the current `StructLiteral` into a new two s.t. `TypedStructLiteral` and `TypelessStructLiteral` for clarity.

Reply:
LGTM for splitting.

In ZetaSQL, both of them are treated in [`ASTStructConstructorWithKeyword`](https://github.com/google/zetasql/blob/194cd32b5d766d60e3ca442651d792c7fe54ea74/zetasql/parser/gen_parse_tree.py#L2757), but they can't be mixed.
* In [typed struct syntax](https://cloud.google.com/spanner/docs/reference/standard-sql/data-types#struct_type)
    > `AS alias` is not allowed on the input expressions.
     > STRUCT<x int64>(5 AS x) | Error - Typed syntax does not allow AS
     > -- | --